internal/credentials/xds/handshake_info_test.go

```diff
@@ -300,3 +312,102 @@ func TestMatchingSANExists_Success(t *testing.T) {
 func newStringP(s string) *string {
 	return &s
 }
+
+func TestEqual(t *testing.T) {
+	mockProvider1 := &mockCertProvider{id: 1}
+	mockProvider2 := &mockCertProvider{id: 2}
+
+	tests := []struct {
+		desc      string
+		hi1       *HandshakeInfo
+		hi2       *HandshakeInfo
+		wantMatch bool
+	}{
+		{
+			desc:      "both HandshakeInfo are nil",
+			hi1:       nil,
+			hi2:       nil,
+			wantMatch: true,
+		},
+		{
+			desc:      "one HandshakeInfo is nil",
+			hi1:       nil,
+			hi2:       NewHandshakeInfo(mockProvider1, nil, nil, false),
+			wantMatch: false,
+		},
+		{
+			desc:      "different root providers",
+			hi1:       NewHandshakeInfo(mockProvider1, nil, nil, false),
+			hi2:       NewHandshakeInfo(mockProvider2, nil, nil, false),
+			wantMatch: false,
+		},
+		{
+			desc:      "different identity providers",
+			hi1:       NewHandshakeInfo(nil, mockProvider1, nil, false),
+			hi2:       NewHandshakeInfo(nil, mockProvider2, nil, false),
+			wantMatch: false,
+		},
+		{
+			desc: "same providers, same SAN matchers",
+			hi1: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("foo.com"), nil, nil, nil, nil, false),
+			}, false),
+			hi2: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("foo.com"), nil, nil, nil, nil, false),
+			}, false),
+			wantMatch: true,
+		},
+		{
+			desc: "same providers, different SAN matchers",
+			hi1: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("foo.com"), nil, nil, nil, nil, false),
+			}, false),
+			hi2: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("bar.com"), nil, nil, nil, nil, false),
+			}, false),
+			wantMatch: false,
+		},
+		{
+			desc: "same SAN matchers with different content",
+			hi1: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("foo.com"), nil, nil, nil, nil, false),
+			}, false),
+			hi2: NewHandshakeInfo(mockProvider1, mockProvider1, []matcher.StringMatcher{
+				matcher.StringMatcherForTesting(newStringP("foo.com"), nil, nil, nil, nil, false),
+				matcher.StringMatcherForTesting(newStringP("bar.com"), nil, nil, nil, nil, false),
+			}, false),
+			wantMatch: false,
+		},
+		{
+			desc:      "different requireClientCert flags",
+			hi1:       NewHandshakeInfo(mockProvider1, mockProvider1, nil, true),
+			hi2:       NewHandshakeInfo(mockProvider1, mockProvider1, nil, false),
+			wantMatch: false,
+		},
+		{
+			desc: "same rootProvider but different mockCertProvider state",
+			hi1: &HandshakeInfo{
+				rootProvider:      mockProvider1,
```

Comment:
Technically, the `HandshakeInfo.Equal` method only compares pointers of the certificate providers. See this line:

```golang
	if hi.rootProvider != other.rootProvider ||
		hi.identityProvider != other.identityProvider ||
		hi.requireClientCert != other.requireClientCert ||
		len(hi.sanMatchers) != len(other.sanMatchers) {
		return false
	}
```

So, I don't think we need to have an `id` field within the `testCertProvider` and we can get rid of any test cases which check that.